Align Vertex Array Objects lifecycle with the spec

lexaknyazev opened the pull request:
#### b2204823ada4999c746c6ef862b72c39bfb20c54
<pre>
Align Vertex Array Objects lifecycle with the spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=242573">https://bugs.webkit.org/show_bug.cgi?id=242573</a>

Reviewed by Kenneth Russell.

* Added checks to deleteVertexArrayOES and isVertexArrayOES
to early exit if the passed object has already been deleted
or created by another context.
* Simplified bindVertexArrayOES by using an existing helper.
* Added pending context checks to WebGL 2.0 versions of
createVertexArray, deleteVertexArray, and isVertexArray.
* Aligned deleteVertexArray error message with the rest of
the codebase.
* Simplified default VAO binding calls when a VAO is
deleted.

* Source/WebCore/html/canvas/OESVertexArrayObject.cpp:
(WebCore::OESVertexArrayObject::deleteVertexArrayOES):
(WebCore::OESVertexArrayObject::isVertexArrayOES):
(WebCore::OESVertexArrayObject::bindVertexArrayOES):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::createVertexArray):
(WebCore::WebGL2RenderingContext::deleteVertexArray):
(WebCore::WebGL2RenderingContext::isVertexArray):

Canonical link: <a href="https://commits.webkit.org/252964@main">https://commits.webkit.org/252964@main</a>
</pre>

kenrussell commented on Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
```diff
@@ -2497,29 +2497,29 @@ void WebGL2RenderingContext::deleteVertexArray(WebGLVertexArrayObject* arrayObje
 
     if (!arrayObject->validate(contextGroup(), *this)) {
-        synthesizeGLError(GraphicsContextGL::INVALID_OPERATION, "deleteVertexArray", "object does not belong to this context");
+        synthesizeGLError(GraphicsContextGL::INVALID_OPERATION, "delete", "object does not belong to this context");
```
I suggest undoing this change. It's helpful to have the precise method name in the console logs.

lexaknyazev replied:
Out of all `delete*` methods, it's the only one that fully spells its name in case of invalid context error. I'll do a follow up to improve reporting throughout.

kenrussell reviewed (approved):
I pointed to your PR https://github.com/KhronosGroup/WebGL/pull/3458 from https://bugs.webkit.org/show_bug.cgi?id=242573 to make the association clear. Please do this in the future - it saves people time in chasing down the references.

Looks good to me.

kenrussell reviewed (approved):
Thanks, still looks good. Trying merge-queue again.

lexaknyazev committed:
Align Vertex Array Objects lifecycle with the spec
https://bugs.webkit.org/show_bug.cgi?id=242573

Reviewed by Kenneth Russell.

* Added checks to deleteVertexArrayOES and isVertexArrayOES
to early exit if the passed object has already been deleted
or created by another context.
* Simplified bindVertexArrayOES by using an existing helper.
* Added pending context checks to WebGL 2.0 versions of
createVertexArray, deleteVertexArray, and isVertexArray.
* Aligned deleteVertexArray error message with the rest of
the codebase.
* Simplified default VAO binding calls when a VAO is
deleted.

* Source/WebCore/html/canvas/OESVertexArrayObject.cpp:
(WebCore::OESVertexArrayObject::deleteVertexArrayOES):
(WebCore::OESVertexArrayObject::isVertexArrayOES):
(WebCore::OESVertexArrayObject::bindVertexArrayOES):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::createVertexArray):
(WebCore::WebGL2RenderingContext::deleteVertexArray):
(WebCore::WebGL2RenderingContext::isVertexArray):

Canonical link: https://commits.webkit.org/252964@main
